Name: Perez, Cris John<br>
Course and Section: CPE 019 - CPE32S9<br>
Date of Submission: March 20, 2024<br>
Instructor: Engr. Roman Richard<br>

###**Hands-on Activity 5.2: Build and Apply Multilayer Perceptron**

Objective: <br>
Build a multilayer perceptron model with the following requirements:


*   Choose any dataset
*   Explain the problem that being solved
*   Create the model
*   Evaluate the accuracy of the model



Preparing the dataset

In [29]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

Import the dataset from uci machine learning repository



In [30]:
!pip install ucimlrepo

In [31]:
from ucimlrepo import fetch_ucirepo

# fetch dataset
rice_cammeo_and_osmancik = fetch_ucirepo(id=545)

# data (as pandas dataframes)
X = rice_cammeo_and_osmancik.data.features
y = rice_cammeo_and_osmancik.data.targets

# metadata
print(rice_cammeo_and_osmancik.metadata)

# variable information
print(rice_cammeo_and_osmancik.variables)

{'uci_id': 545, 'name': 'Rice (Cammeo and Osmancik)', 'repository_url': 'https://archive.ics.uci.edu/dataset/545/rice+cammeo+and+osmancik', 'data_url': 'https://archive.ics.uci.edu/static/public/545/data.csv', 'abstract': "A total of 3810 rice grain's images were taken for the two species, processed and feature inferences were made. 7 morphological features were obtained for each grain of rice.", 'area': 'Biology', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 3810, 'num_features': 7, 'feature_types': ['Real'], 'demographics': [], 'target_col': ['Class'], 'index_col': None, 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 2019, 'last_updated': 'Fri Nov 03 2023', 'dataset_doi': '10.24432/C5MW4Z', 'creators': [], 'intro_paper': {'title': 'Classification of Rice Varieties Using Artificial Intelligence Methods', 'authors': 'Ilkay Cinar, M. Koklu', 'published_in': 'International Journal of Intelligent Systems and App

In [47]:
#Encoding categorical labels into numerical values
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)

/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:116: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [48]:
y

array([0, 0, 0, ..., 1, 1, 1])

In [44]:
#Standardizing features
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X = scaler.fit_transform(X)

In [45]:
X

array([[ 1.47982953,  2.0043543 ,  2.34854657, ...,  2.01833746,
         1.49965944, -1.15292093],
       [ 1.14787029,  1.12585309,  0.98839042, ...,  0.41001813,
         1.19291767, -0.60207876],
       [ 1.13516924,  1.31721425,  1.45190846, ...,  1.21295648,
         1.12650386,  0.405611  ],
       ...,
       [-0.13320373, -0.32985087, -0.29824512, ..., -0.27509915,
        -0.17306812, -0.45573108],
       [-1.60825742, -1.74032002, -1.58097116, ..., -0.59882135,
        -1.60715621, -0.03716757],
       [-0.71225612, -1.39156604, -1.58754648, ..., -2.93916012,
        -0.76628981,  1.82594693]])

The dataset that I used was the rice dataset (cammeo and osmancik) wherein there are a total of 3810 sammples of ruce grain's images that would classify them if they are cammeo or osmancik.

The problem that I am trying to solve is to classify rice grains into one of the two types based on their dimensions or measurements.



Creation of the model

In [49]:
#Split the dataset into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33, random_state = 42)

In [53]:
print("Feature matrix:", X_train.shape)
print("Target matrix:", X_test.shape)
print("Feature matrix:", y_train.shape)
print("Target matrix:", y_test.shape)

Feature matrix: (2552, 7)
Target matrix: (1258, 7)
Feature matrix: (2552,)
Target matrix: (1258,)


In [50]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

model = Sequential()
model.add(Dense(10, input_dim = X.shape[1], activation = 'relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(len(np.unique(y)),activation = 'softmax'))

In [65]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 10)                80        
                                                                 
 dense_1 (Dense)             (None, 8)                 88        
                                                                 
 dense_2 (Dense)             (None, 2)                 18        
                                                                 
Total params: 186 (744.00 Byte)
Trainable params: 186 (744.00 Byte)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [51]:
#Compile the model
model.compile(loss='sparse_categorical_crossentropy',optimizer='adam',metrics=['accuracy'])

In [70]:
#Fit the model
model.fit(X_train, y_train, epochs=10, batch_size=3810, validation_split=0.2)

Epoch 1/10
1/1 [==============================] - 0s 91ms/step - loss: 0.2073 - accuracy: 0.9201 - val_loss: 0.2008 - val_accuracy: 0.9217
Epoch 2/10
1/1 [==============================] - 0s 40ms/step - loss: 0.2071 - accuracy: 0.9206 - val_loss: 0.2006 - val_accuracy: 0.9217
Epoch 3/10
1/1 [==============================] - 0s 37ms/step - loss: 0.2069 - accuracy: 0.9201 - val_loss: 0.2003 - val_accuracy: 0.9217
Epoch 4/10
1/1 [==============================] - 0s 56ms/step - loss: 0.2067 - accuracy: 0.9206 - val_loss: 0.2000 - val_accuracy: 0.9217
Epoch 5/10
1/1 [==============================] - 0s 42ms/step - loss: 0.2065 - accuracy: 0.9206 - val_loss: 0.1998 - val_accuracy: 0.9217
Epoch 6/10
1/1 [==============================] - 0s 55ms/step - loss: 0.2062 - accuracy: 0.9206 - val_loss: 0.1995 - val_accuracy: 0.9217
Epoch 7/10
1/1 [==============================] - 0s 56ms/step - loss: 0.2060 - accuracy: 0.9206 - val_loss: 0.1992 - val_accuracy: 0.9217
Epoch 8/10
1/1 [===========

Evaluation of the accuracy of the model

In [71]:
results = model.evaluate(X_test, y_test, verbose = 1)
print('test loss, test acc:',results)

40/40 [==============================] - 0s 2ms/step - loss: 0.1928 - accuracy: 0.9316
test loss, test acc: [0.19281327724456787, 0.9316375255584717]


By using the ```model.evaluate()``` on the test samples, the accuracy of the model was 93%.




Dataset used: Rice (Cammeo and Osmancik). (2019). UCI Machine Learning Repository. https://doi.org/10.24432/C5MW4Z.